### ISSUE TITLE
Relaxing inferenced shape

### ISSUE THREAD

**Issue by gaohuazuo** [tensorflow/tensorflow#6906]
Inferenced shape will put restriction on feeding, so sometimes it is desirable to relax inferenced shape. Currently this can be done by `lambda x: tf.cond(tf.constant(True), lambda: x, lambda: tf.placeholder(x.dtype))`, or `tf.placeholder_with_default` if gradient is not required (`placeholder_with_default` has no gradient registered).

I also read a comment in `identity_op.cc`: 

>PlaceholderWithDefault does the same thing as Identity, but has a
>different shape function (and constant value function) registered.

Possibly register a gradient for `placeholder_with_default`?

**Comment by asimshankar** [comment]
@vrv : Any thoughts here? I'm not quite sure what the position is on feeding arbitrary nodes in the graph as it can can mess with shape inference, gradient computations (see #6903 that @gaohuazuo also filed).

**Comment by vrv** [comment]
I have no thoughts because I don't really understand the problem.  Specifying a placeholder without shape allows feeding any shape data.

I don't think we register gradients for placeholders either.  I think providing more context into what you're trying to do will help us out.

**Comment by yaroslavvb** [comment]
I don't understand the problem either, but you can specify the gradient for placeholder_with_default using `override_gradient_map` (it'll return `None` instead of crashing with error on gradient computation)

**Comment by gaohuazuo** [comment]
Thanks for the replies. What I'm doing is to explore the behaviour of the discriminator in GAN. The graph is roughly the following:

![untitled diagram](https://cloud.githubusercontent.com/assets/10446514/22094828/4b0cfaf0-de4b-11e6-97e3-f3c6b263b04d.png)

During training, gradients have to propagate through `x` to `G`. On the other hand, I want to feed some arbitrary `x` to `D` for testing. I feel that this should be suitable for `placeholder_with_default`, but it has no gradient and thereby prevents training using the same graph. Note that I either feed `x` or use the gradient, but never both, so this should be a good use case of `placeholder_with_default`.

I guess it is a protective measure that `placeholder_with_default` doesn't have a gradient. However, this isn't quite meaningful, as other ops don't have the same protection, and `placeholder_with_default` has nothing special to do with feeding, except its name.

In short, I would suggest using `identity` as the gradient of `placeholder_with_default`, because it does no harm and some good.

**Comment by vrv** [comment]
You should be able to just feed whatever you want to D during testing (and presumably since you don't need the gradient during testing, you're not fetching anything below D either).  Doing so will induce a subgraph that cuts out everything before it.

**Comment by vrv** [comment]
Hm, connecting this back to now what I think your original request is, that you want to avoid having the input to 'x' be a well known shape, I think I understand why you want this.  It seems plausible to add the identity gradient for placeholder_with_default.

cc @mrry who added placeholder_with_default

It might also be possible to set the shape of 'x' in python using set_shape() to something that has an unknown shape before connecting it to D.

**Comment by mrry** [comment]
Registering `tf.identity()` as the gradient of `tf.placeholder_with_default()` makes sense to me as a first step.

The only downside&mdash;as @gaohuazuo implied&mdash;is that if you feed a value for a `tf.placeholder_with_default()` **and** consume its gradient, the gradient will depend on the default value, which might run counter to expectations. I'm open to suggestions on what the right behavior should be here.

**Comment by yaroslavvb** [comment]
backprop op for`tf.identity` propagates backprops unchanged, so in this sense it doesn't depend on input value

**Comment by yaroslavvb** [comment]
Actually, it seems the author is requesting that `tf.placeholder_with_default` has 0 gradient rather than gradient of identity

**Comment by gaohuazuo** [comment]
Let me summarize a bit:
1. Registering `identity` as the gradient for `placeholder_with_default` solves this issue perfectly.
2. There are problems related to feeding, but they are not specific to `placeholder_with_default` nor its gradient, and should be addressed separately in #6903.
3. I guess we can register `identity` for the gradient and close this issue.

P.S I originally thought that a 'good' gradient for `placeholder_with_default` would solve this issue and #6903, but that's not the case (see #6903).

**Comment by vrv** [comment]
@gaohuazuo: Thanks for the summary.  Given that as @yaroslavvb mentioned, the gradient is independent of the fed input (it's not clear how the gradient should be different than identity when fed), I think it seems reasonable to add the gradient.  Would you like to send a PR for this with a test to validate your expectations?
